src/Features/CSharp/Portable/PullMemberUp/CSharpMembersPullerService.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.  
+
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp
+{
+    [ExportLanguageService(typeof(IMembersPullerService), LanguageNames.CSharp), Shared]
+    internal class CSharpMembersPullerService : AbstractMembersPullerService<UsingDirectiveSyntax>
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpMembersPullerService()
+        {
+        }
+
+        protected override SyntaxTriviaList GetFirstMemberTrivia(SyntaxNode root)
+        {
+            var members = root switch
+            {
+                CompilationUnitSyntax compilationUnit => compilationUnit.Members,
+                NamespaceDeclarationSyntax namespaceDeclaration => namespaceDeclaration.Members,
+                _ => throw ExceptionUtilities.UnexpectedValue(root)
+            };
+
+            if (members.Count == 0)
+            {
+                return SyntaxTriviaList.Empty;
+            }
+
+            return members.First().GetLeadingTrivia();
+        }
+
+        protected override SyntaxNode EnsureLeadingTriviaBeforeFirstMember(SyntaxNode root, SyntaxTriviaList trivia)
+        {
+            var members = root switch
+            {
+                CompilationUnitSyntax compilationUnit => compilationUnit.Members,
+                NamespaceDeclarationSyntax namespaceDeclaration => namespaceDeclaration.Members,
+                _ => throw ExceptionUtilities.UnexpectedValue(root)
+            };
+
+            if (members.Count == 0)
+            {
+                return root;
+            }
+
+            var firstMember = members.First();
+            return root.ReplaceNode(firstMember, firstMember.WithLeadingTrivia(trivia));
+        }
+
+        protected override async Task<IEnumerable<UsingDirectiveSyntax>> GetImportsAsync(Document document, CancellationToken cancellationToken)
```

Comment:
If you are not trying to optimize the perf by using IEnumerable, then using ImmutableArray instead as the return type.
It would prevent the possible enumerations.

Reply:
I think I understand. Should I be changing the `IsValidUnnecessaryImport` method to accept an `ImmutableArray` as well?

Reply:
Yes, feel free to do that. 
P.S. When you change this, in Roslyn there are a set of utilities, like SelectAsArray(). You could use this to avoid call ToImmutableArray() a lot of times.